### PR TITLE
get extended characteristics and print cet compat

### PR DESCRIPTION
**Pull request by yardenshafir**
This PR relies on the newest release of pe-parse (v2.1.0). There is an open PR in vcpkg to update to the latest version: https://github.com/microsoft/vcpkg/pull/27811

### CODE REVIEW

File: checksec.cpp [hunk 1]
```diff
@@ -63,6 +64,35 @@ Checksec::Checksec(std::string filepath) : filepath_(filepath), loadedImage_(fil
...
+        peparse::debug_dir_entry debugDir{};
+        uint32_t numberOfDebugDirs = debugDirectories.size() / sizeof(debugDir);
+        auto debugDirSize = std::min(debugDirectories.size(), sizeof(debugDir));
+
+        for (int i = 0; i < numberOfDebugDirs; i++) {
+            // copy the current debug directory into debugDir
+            memcpy(&debugDir, debugDirectories.data() + (i * debugDirSize), debugDirSize);
+            // 20 == IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS
+            // For now we only care about this debug directory since it contains the CETCOMPAT bit
+            if (debugDir.Type == 20) {
```
Comment (woodruffw) [inline checksec.cpp:85]: Can we use the `IMAGE_DLLCHARACTERISTICS_EX_CET_COMPAT` constant from `pe-parse` here?

File: checksec.cpp [hunk 2]
```diff
@@ -63,6 +64,35 @@ Checksec::Checksec(std::string filepath) : filepath_(filepath), loadedImage_(fil
...
+                if (debugDir.PointerToRawData != 0) {
+                    auto dataPtr = debugDir.PointerToRawData + loadedImage_.get()->fileBuffer->buf;
+                    try {
+                        extendedDllCharacteristics_ = *((uint16_t*)dataPtr);
+                    } catch (...) {
+                        extendedDllCharacteristics_ = 0;
+                    }
```
Comment (woodruffw) [inline checksec.cpp:92]: I might be wrong, but I think this `try...catch` isn't doing anything -- the only thing that can really happen here is a fault if `dataPtr` isn't a valid address, but `catch` won't handle that. 

I think instead we want to make sure `dataPtr` is valid within the `fileBuffer`, and remove the block.

File: test/winchecksec-test.cpp [hunk 3]
```diff
@@ -149,3 +149,11 @@ TEST(Winchecksec, CFG64) {
...
     }
 }
+
+TEST(Winchecksec, NoCetCompat64) {
+    auto *path = WINCHECKSEC_TEST_ASSETS "/64/pegoat-no-cetcompat.exe";
+
+    auto checksec = checksec::Checksec(path);
+
+    EXPECT_FALSE(checksec.isCetCompat());
+}
```
Comment (woodruffw) [inline test/winchecksec-test.cpp:159]: N.B.: We might want to add another "goat" binary; the repo is here: https://github.com/trailofbits/pegoat

File: checksec.cpp (hunk 1 above)
Comment (yardenshafir) [reply checksec.cpp:85]: Looks like I never defined it as a constant in `pe-parse` :(

File: checksec.cpp (hunk 2 above)
Comment (yardenshafir) [reply checksec.cpp]: Fixed!

File: test/winchecksec-test.cpp (hunk 3 above)
Comment (yardenshafir) [reply test/winchecksec-test.cpp:159]: Added testers :)

Comment (woodruffw) [reply test/winchecksec-test.cpp:159]: Thanks! Could you add test stubs for them, just ensuring `EXPECT_TRUE(checksec.isCetCompat())`?

File: checksec.cpp (hunk 2 above)
Comment (woodruffw) [reply checksec.cpp]: LGTM, thanks!

File: test/winchecksec-test.cpp (hunk 3 above)
Comment (woodruffw) [reply test/winchecksec-test.cpp:159]: Hmm, I just noticed these new binaries aren't in the `pegoat` repo's build. Did you build them locally?

Comment (woodruffw) [reply test/winchecksec-test.cpp:159]: Thanks!

**Review by woodruffw — APPROVED** [review]
LGTM!

**Pull request merged**